DOC: cleanup - containers, organization, etc. (#193)
pcdshub/happi: docs/source/containers.rst

* Add default values when unspecified
* Enforce - or validate - a certain format for the field

.. autoclass:: happi.EntryInfo
   :members:

HappiItem
^^^^^^^^^

In order to ensure that information is entered into the database in an
organized fashion, the client will only accept classes that inherit from
:class:`.HappiItem`. Each item will have the key information represented as
class attributes, available to be manipulated like any other regular property

Editing the information for a container is a simple as:

.. ipython:: python

    from happi import HappiItem

    item = HappiItem(name='my_device')

    item.name = 'new_name'

.. note::

    :class:`happi.Device` class is **deprecated** due to ambiguous name,
    conflicting with :class:`ophyd.Device`.

Example Container
^^^^^^^^^^^^^^^^^

In order to show the flexibility of the :class:`.EntryInfo` class, we'll put
together a new example container. The class can be invoked in the same way you
would usually handle class inheritance, the only difference is that you specify
class attributes as EntryInfo objects:

.. ipython:: python

    import re
    from happi import HappiItem, EntryInfo

>     class MyDevice(HappiItem):
>         """My new device, with a known model number."""
>         model_no      = EntryInfo('Model Number of Device', optional=False)
>         count         = EntryInfo('Count of Device', enforce=int, default=0)

Reviewer: We should further purge the `Device` keyword to avoid confusion/bad habits:
```suggestion
    class MyItem(HappiItem):
        """My new item, with a known model number."""
        model_no      = EntryInfo('Model Number of Item', optional=False)
        count         = EntryInfo('Count of Item', enforce=int, default=0)
```

        choices       = EntryInfo('Choice Info', enforce=['a','b','c'])
        no_whitespace = EntryInfo('Enforce no whitespace',
                                   enforce = re.compile(r'[\S]*$'))

> By default, `.EntryInfo` will create an optional init keyword argument with a

Reviewer: Does this need a class directive?
```suggestion
By default, :class:`.EntryInfo` will create an optional init keyword argument with a
```

default of ``None`` with the same name as the class attribute. A quick way
to see how this information will be put into the the database is taking a look

> at ``dict(device)``:
> 
> .. ipython:: python
> 
>     device = MyDevice(name="my_device", model_no="QABC1234")
>     dict(device)

Reviewer: ```suggestion
at ``dict(item)``:

.. ipython:: python

    item = MyItem(name="my_item", model_no="QABC1234")
    dict(item)
```

As shown in the example above, using the EntryInfo keywords, you can put a
short doc string to give a better explanation of the
field, and also enforce that user enter a specific format of information.

While the user will always be able to enter ``None`` for the attribute, if a
real value is given it will be checked against the specified ``enforce``
keyword, raising ``ValueError`` if invalid. Here is a table for how the
:class:`.EntryInfo` check the type

=======   ===========================
Enforce   Method of enforcement
=======   ===========================
None      Any value will work
type      type(value)
list      list.index(value)
regex     regex.match(value) != None
=======   ===========================

Fields that are important to the device can be marked as mandatory
(``optional=False``); these should have no default value.

> When entering information you will not neccesarily see a difference in between
> optional and mandatory :class:`.EntryInfo`, however the database client will

Reviewer: ```suggestion
When entering information you will not neccesarily see a difference between
optional and mandatory :class:`.EntryInfo`, however the database client will
```

reject the device if these fields do not have the requisite values set.

Loading your Object
^^^^^^^^^^^^^^^^^^^

A container's primary role is containing the information necessary to load the
Python representation of an object.

Internally, happi keeps track of all device containers by way of its registry,
the :class:`~happi.containers.HappiRegistry`.

This information is stored as a ``device_class``, ``args`` and ``kwargs``. The
former stores a string that indicates the Python class of the device, the other
two indicate the information that is needed to instantiate it. With this
information both :func:`.from_container` and :func:`.load_device` will handle
importing modules and instantiating your object.

.. note::

    happi will attach the original metadata with a fixed attribute name ``.md``
    to your object.  You can use this to keep track of the container metadata
    used to instantiate your object.

    This can be disabled by setting ``attach_md`` to ``False`` in

>     `from_container`.

Reviewer: Does this need a func directive?
```suggestion
    :func:`from_container`.
```

Often information contained in the ``args`` or ``kwargs`` will be duplicated in
other parts of the container. For instance most ``ophyd`` objects will want a
``name`` and ``prefix`` on initialization. Instead of repeating that
information you can just use a template and have the information automatically
populated for you by the container itself. For instance, in the aforementioned
example ``device.args = ["{{prefix}}"]`` would substitute the attribute listed as
``Device.prefix`` in as an argument. If the template contains the substituted
attribute alone the type will also be converted.

.. ipython:: python

    from happi import from_container

>     container = MyDevice(name="my_device", model_no="QABC1234",

Reviewer: ```suggestion
    container = MyItem(name="my_item", model_no="QABC1234",
```

Author: Fixed up while rebuilding after accepting the changes. Thanks!

Then, add an
`entry point <https://packaging.python.org/specifications/entry-points/>`_
specified by the **happi.containers** keyword to your package ``setup.py``.
An example entry points can be found `here
<https://github.com/pcdshub/pcdsdevices/blob/master/setup.py>`_.

:class:`~happi.containers.HappiRegistry` takes care of loading the entry points
and making them available throughout the library.

.. _convention_label:

Built-in Container Conventions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order for the database to be as easy to parse as possible we need to
establish conventions on how information is entered. Please read through this
before entering any information into the database.

HappiItem Entries
+++++++++++++++++

These are fields that are common to all Happi items.

**name**

This is simply a short name we can use to refer to the device.

**device_class**

This is the full class name, which lets happi know how to instantiate your
device.

.. note::

    This may also be the name of a factory function - happi only cares that
    it's callable.

**args**

Argument list to be passed on instantiation.  May contain templated macros such

> as ``{{name}}``.

Reviewer: Making the example more correct for args in the ophyd context
```suggestion
as ``{{prefix}}``.
```

"entries" of type :class:`.EntryInfo`.

EntryInfo
^^^^^^^^^

In order to regulate and template the information that gets entered into the
Happi database, we use the concept of containers. This allows a developer to
specify fields that are essential to every instance of a specific type.

These fields are specified using an instance of :class:`.EntryInfo`.
`.EntryInfo` provides several primary features:

* Mark a field as required (or optional)
* Add default values when unspecified
* Enforce - or validate - a certain format for the field

.. autoclass:: happi.EntryInfo
   :members:

HappiItem
^^^^^^^^^

In order to ensure that information is entered into the database in an
organized fashion, the client will only accept classes that inherit from
:class:`.HappiItem`. Each item will have the key information represented as
class attributes, available to be manipulated like any other regular property

Editing the information for a container is a simple as:

.. ipython:: python

    from happi import HappiItem

    item = HappiItem(name='my_device')

    item.name = 'new_name'

.. note::

    :class:`happi.Device` class is **deprecated** due to ambiguous name,

>     conflicting with :class:`ophyd.Device`.

Author: Yes, `Device` is deprecated but not removed.

together a new example container. The class can be invoked in the same way you
would usually handle class inheritance, the only difference is that you specify
class attributes as EntryInfo objects:

.. ipython:: python

    import re
    from happi import HappiItem, EntryInfo

    class MyItem(HappiItem):
        """My new item, with a known model number."""
        model_no      = EntryInfo('Model Number of Item', optional=False)
        count         = EntryInfo('Count of Item', enforce=int, default=0)
        choices       = EntryInfo('Choice Info', enforce=['a','b','c'])
        no_whitespace = EntryInfo('Enforce no whitespace',
                                   enforce = re.compile(r'[\S]*$'))

By default, :class:`.EntryInfo` will create an optional init keyword argument with a
default of ``None`` with the same name as the class attribute. A quick way
to see how this information will be put into the the database is taking a look
at ``dict(item)``:

.. ipython:: python

    item = MyItem(name="my_item", model_no="QABC1234")
    dict(item)

As shown in the example above, using the EntryInfo keywords, you can put a

> short doc string to give a better explanation of the

Reviewer 2: ```suggestion
short docstring to give a better explanation of the
```

Reviewer 2: Out of curiosity, how is this docstring ever really visible?
I kind of see some `make_docstring()` and such handling stuff but I'm tired and can't find my docstring anywhere :/

Author: I think the answer is probably "not yet". It'd be good to generate field documentation down the line with this, similar to how we use it in ophyd/caproto documentation though.

field, and also enforce that user enter a specific format of information.

While the user will always be able to enter ``None`` for the attribute, if a
real value is given it will be checked against the specified ``enforce``
keyword, raising ``ValueError`` if invalid. Here is a table for how the
:class:`.EntryInfo` check the type

=======   ===========================
Enforce   Method of enforcement
=======   ===========================
None      Any value will work

> type      type(value)

Author: I think since we deal mostly with strings in the command-line, the enforcement is geared toward that. That would indicate `int(string_value)` instead of `type(string_value) is int`.

This copied-in documentation could definitely use some rewording though.

list      list.index(value)
regex     regex.match(value) != None
=======   ===========================

> Fields that are important to the item can be marked as mandatory
> (``optional=False``); these should have no default value.

Reviewer 2: ```suggestion
Fields that are important to the item can be marked as mandatory with
``optional=False`` and should have no default value.
```

When entering information you will not neccesarily see a difference between
optional and mandatory :class:`.EntryInfo`, however the database client will

> reject the item if these fields do not have the requisite values set.

Reviewer 2: Not a docs issue, but I feel like there should be some kind of warning if you instantiate a container without required data, no?

Author: I think the original intention was that you can create it and then modify it in-place.

I think it's not so crazy to ultimately be rejected by the database for a malformed entry - it'd indicate your container's field `enforce` settings were incorrect to let it get to that point, I guess?

Loading your Object
^^^^^^^^^^^^^^^^^^^

A container's primary role is containing the information necessary to load the
Python representation of an object.

Internally, happi keeps track of all containers by way of its registry, the
:class:`~happi.containers.HappiRegistry`.

This information is stored as a ``device_class``, ``args`` and ``kwargs``. The
former stores a string that indicates the Python class of the item, the other
two indicate the information that is needed to instantiate it. With this

> information both :func:`.from_container` and :func:`.load_device` will handle
> importing modules and instantiating your object.

Reviewer 2: ```suggestion
information both :func:`.from_container` and :func:`.load_device` are able to
handle importing modules and instantiating your object.
```

.. note::

    happi will attach the original metadata with a fixed attribute name ``.md``
    to your object.  You can use this to keep track of the container metadata
    used to instantiate your object.

    This can be disabled by setting ``attach_md`` to ``False`` in
    :func:`.from_container`.

Often information contained in the ``args`` or ``kwargs`` will be duplicated in
other parts of the container. For instance most ``ophyd`` objects will want a

> ``name`` and ``prefix`` on initialization. Instead of repeating that

Reviewer 2: `name` and `prefix` typically aren't the same value but I can't think of a better example so 👍

> Often information contained in the ``args`` or ``kwargs`` will be duplicated in
> other parts of the container. For instance most ``ophyd`` objects will want a
> ``name`` and ``prefix`` on initialization. Instead of repeating that
> information you can just use a template and have the information automatically
> populated for you by the container itself. For instance, in the aforementioned
> example ``container.args = ["{{name}}"]`` would substitute the attribute
> listed as ``container.name`` in as an argument. If the template contains the
> substituted attribute alone the type will also be converted.

Reviewer 2: ```suggestion
Often, information contained in the ``args`` or ``kwargs`` will be duplicated
in other parts of the container. For instance, most ``ophyd`` objects will want
a ``name`` and ``prefix`` on initialization. Instead of repeating that
information you can just use a template and have the information automatically
populated for you by the container itself. For instance, in the aforementioned
example, ``container.args = ["{{name}}"]`` would substitute the value of
``container.name`` in as an argument. If the template contains the substituted
attribute alone, the type will also be converted.
```

.. ipython:: python

    from happi import from_container

    container = MyItem(name="my_item", model_no="QABC1234",
                       device_class='ophyd.sim.SynSignal',
                       kwargs={'name': '{{name}}'})
    obj = from_container(container, attach_md=True)
    obj

Integrating with your package
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Happi provides some containers for your convenience, but intentionally does
not support all use cases or control systems in a monolithic fashion.

The suggested method for supporting your new package would be to make your

> package have a dependency on happi, of course, and subclass `.HappiItem`

Reviewer 2: ```suggestion
package have a dependency on happi, of course, and subclass :class:`.HappiItem`
```
The other option is to add
```
# The reST default role (used for this markup: `text`)
default_role = 'any'
```
to the conf.py and get rid of the `.` here but that raises a couple other warnings that would need to be fixed.

to make a new container in your own package.

Then, add an
`entry point <https://packaging.python.org/specifications/entry-points/>`_

> specified by the **happi.containers** keyword to your package ``setup.py``.
> An example entry points can be found `here

Reviewer 2: ```suggestion
specified by the **happi.containers** keyword to your package's ``setup.py``.
Example entry points can be found `here
```

<https://github.com/pcdshub/pcdsdevices/blob/master/setup.py>`_.

:class:`~happi.containers.HappiRegistry` takes care of loading the entry points
and making them available throughout the library.

.. _convention_label:

Built-in Container Conventions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order for the database to be as easy to parse as possible we need to
establish conventions on how information is entered. Please read through this
before entering any information into the database.

> HappiItem Entries

Reviewer 2: _maybe_ backtick the class name... _maybe_...
It's more helpful but it does look kinda ugly with the box around it in the heading

+++++++++++++++++

These are fields that are common to all Happi items.

> **name**

Reviewer 2: I feel like all of these fields and their descriptions should be indented. What do you think?

Author: Sure 👍

This is simply a short name we can use to refer to the item.

**device_class**

This is the full class name, which lets happi know how to instantiate your
item.

The ``device_class`` name remains for backward-compatibility reasons.  Thinking
of it as ``class_name`` or ``creator_callable`` would be more apt.

.. note::

    This may also be the name of a factory function - happi only cares that
    it's callable.

**args**

Argument list to be passed on instantiation.  May contain templated macros such
as ``{{name}}``.

**kwargs**

Keyword argument dictionary to be passed on instantiation.  May contain
templated macros such as ``{{name}}``.

**active**

Is the object actively deployed?

**documentation**

A brief set of information about the object.

OphydItem entries
+++++++++++++++++

`ophyd <blueskyproject.io/ophyd/>`_ has first-class support in happi - but
not much is required on top of the base HappiItem to support it.

**prefix**

This should be the prefix for all of the PVs contained within the device. It
does not matter if this is an invalid record by itself.

LCLSItem entries

> ++++++++++++++++

Reviewer 2: Should this section still be here if `LCLSItem` isn't in the package?
